### PR TITLE
[V3]Network fix

### PR DESCRIPTION
**Pull request by PraveenPenguin**
cleanup and some new utilities added

Signed-off-by: Praveen K Pandey praveen@linux.vnet.ibm.com

### CODE REVIEW

File: avocado/utils/configure_network.py [hunk 1]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
     """
 
 
-def set_ip(ipaddr, netmask, interface, interface_type=None):
+class NetworkInterface:
     """
-    Gets interface name, IP, subnet mask and creates interface
-    file based on distro.
+    class For network interface
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: We need a better docstring here.

Please, let's try to follow  PEP-0257:

https://www.python.org/dev/peps/pep-0257/

File: avocado/utils/configure_network.py [hunk 2]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
-        except process.CmdError as ex:
-            raise NWException("ifup fails: %s" % ex)
+    def __init__(self, if_name, remote_session=None):  # pylint: disable=W0231
+        self.name = if_name
+        self.remote_session = remote_session
 
+    def set_ip(self, ipaddr, netmask, interface_type=None):
+        """
+        utility assign a IP  address (given to this utility ) to  Interface
+        and gernerate interface file based on distro
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: The same here.

File: avocado/utils/configure_network.py [hunk 3]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
+                network_conf.write("IPV6INIT=yes \n")
+                network_conf.write("IPV6_AUTOCONF=yes \n")
+                network_conf.write("IPV6_DEFROUTE=yes")
+        self.bring_up()
 
-    if distro.detect().name == 'SuSE':
-        conf_file = "/etc/sysconfig/network/ifcfg-%s" % interface
+    def unset_ip(self):
+        """
+        Utility to unassign IP to Defined Interface
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: One-line docstrings should follow the format:
```
"""string here."""
```

File: avocado/utils/configure_network.py [hunk 4]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
 
-    cmd = "ifdown %s" % interface
-    try:
-        process.system(cmd, sudo=True)
-        shutil.move(conf_file+".backup", conf_file)
+        if distro.detect().name == 'SuSE':
+            conf_file = "/etc/sysconfig/network/ifcfg-%s" % self.name
+
+        self.bring_down()
+        self._move_config_file("%s.backup" % conf_file, conf_file)
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: If distro.detect().name is something else, conf_file here will be undefined, right?

File: avocado/utils/configure_network.py [hunk 5]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
+        Utility to unassign IP to Defined Interface
+        """
+        if distro.detect().name == 'rhel':
+            conf_file = "/etc/sysconfig/network-scripts/ifcfg-%s" % self.name
 
-    cmd = "ifdown %s" % interface
-    try:
-        process.system(cmd, sudo=True)
-        shutil.move(conf_file+".backup", conf_file)
+        if distro.detect().name == 'SuSE':
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: You don't need to call `detect()` method every time inside the if. You could avoid multiple calls with:
```python
distro = distro.detect().name
if distro == 'SuSe':
   print('suse')
elif distro == 'rhel':
   print('rhel')
else:
   raise Exception("Distro Unknow, could not set ip")
```

File: avocado/utils/configure_network.py [hunk 6]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
+            conf_file = "/etc/sysconfig/network/ifcfg-%s" % self.name
+
+        self.bring_down()
+        self._move_config_file("%s.backup" % conf_file, conf_file)
+
+    def ping_check(self, peer_ip, count, option=None, flood=False):
+        """
+        utility perform Ping operation on Peer IP address and return status
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: Same here for docstring.

File: avocado/utils/configure_network.py [hunk 7]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
+        if process.system(cmd, shell=True, verbose=True,
+                          ignore_status=True) != 0:
+            return False
         return True
-    except Exception as ex:
-        raise NWException("ifdown fails: %s" % ex)
 
+    def set_mtu(self, mtu):
+        """
+        utility set mtu size to a interface and return status
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: Same here... you got it, right?

File: avocado/utils/configure_network.py [hunk 8]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
+                mtu_value = process.system_output(
+                    cmd_mtu, shell=True).decode("utf-8")
+            if mtu in mtu_value and wait.wait_for(self.is_link_up, timeout=120, args=[self.name]):
+                return True
+        except Exception:  # pylint: disable=W0703
+            return False
         return False
-    return True
 
+    def get_link_status(self):
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: Missing docstring.

File: avocado/utils/configure_network.py [hunk 9]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
+        if self.remote_session:
+            return self.session.cmd("cat /sys/class/net/%s/operstate" % self.name).stdout.decode('utf-8')
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: Why do you need decode here? I did a small test locally and it works without the decode()

File: avocado/utils/configure_network.py [hunk 10]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
+            shutil.move(src_conf, dest_conf)
+        else:
+            raise NWException("%s interface not available" % self.name)
+
+    def get_hwaddr(self):
+        try:
+            interface_file = open(
+                '/sys/class/net/%s/address' % self.name)
+            hwaddr = interface_file.read().strip()
+            interface_file.close()
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]:
```python
with open(filename, 'r') as fp:
   return fp.read().strip()
```

File: avocado/utils/configure_network.py [hunk 11]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
-        create a object for accesses remote machine
+        Utility which set Hw address to Interface
+        :param hwaddr: Pass Hardwae address for defined interface
         """
         try:
-            self.session = Session(host, port=port, user=peer_user,
-                                   key=key, password=peer_password)
-        except Exception as ex:  # pylint: disable=W0703
-            log.error("connection not established to peer machine: %s", ex)
+            process.system('ifconfig %s hw ether %s' %
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: Sometimes you use `ifconfig`, sometimes `ip`. Is there any reason to not use `iproute` here?

```bash
$  ip link set eth0 address 02:01:02:03:04:08
```

File: avocado/utils/configure_network.py [hunk 12]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
+                           (self.name, hwaddr))
+        except Exception as ex:
+            raise NWException("ifdown fails: %s" % ex)
 
-    def set_mtu_peer(self, peer_interface, mtu):
+    def add_hwaddr_address(self, maddr):
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: This is redundant right? `add_hardware_address` or `add_hwadrr`

File: avocado/utils/configure_network.py [hunk 13]
```diff
@@ -35,168 +37,248 @@ class NWException(Exception):
...
-        except Exception as ex:  # pylint: disable=W0703
-            log.error("setting MTU value in peer failed: %s", ex)
-        return False
+            process.system('ip maddr add %s dev %s' % (maddr, self.name))
+            return True
+        except Exception as ex:
+            raise NWException("ifdown fails: %s" % ex)
 
-    def get_peer_interface(self, peer_ip):
+    def remove_hwaddr_address(self, maddr):
```
Comment (beraldoleal) [inline avocado/utils/configure_network.py]: same here.

File: avocado/utils/configure_network.py (hunk 9 above)
Comment (PraveenPenguin) [reply avocado/utils/configure_network.py]: ah yes right it should, will adopt

**Comment by PraveenPenguin** [comment]
closing in favour of #3628, @beraldoleal adopted it request a change  please have a look on that PR

**Pull request closed without merging**